docs/src/index.md

```diff
@@ -170,7 +172,9 @@ julia> dmodel = gradient(model -> sum(model(x)), model)[1]
 (W = [0.652543 … 0.683588], b = [1.0, 1.0])
 ```
 
-Zygote also support one more way to take gradients, via *implicit parameters* – this is a lot like autograd-style gradients, except we don't have to thread the parameter collection through all our code.
+### Implicit parameters
+
+Zygote also support one more way to take gradients, via *implicit parameters* – this is a lot like autograd-style gradients, except we don't have to thread the parameter collection through all our code. When working with Flux models, this is the recommended way of passing the gradients, as it ensures compatibility with Flux's built-in optimizers. 
```

Comment:
```suggestion
Zygote also supports taking gradients via *implicit parameters*. This is a lot like autograd-style gradients, except we don't have to thread the parameter collection through our code.
```
We're moving away from using implicit parameters in general, and the next iteration of Flux's optimizers are designed for explicit parameters/gradients first. I'm also leery of referencing a downstream package, especially when we already have similar documentation in https://fluxml.ai/Flux.jl/stable/models/basics/#Taking-Gradients and https://fluxml.ai/Flux.jl/stable/training/optimisers/. Perhaps an alternative would be to link

Reply:
In the long term, I agree, but the whole point of this PR is to prevent the frustration of reading Zygote's docs, taking explicit gradients, and then getting an error on `update!`. Flux is already mentioned in the docs.

Reply:
Why are people reading Zygote's docs instead of Flux's docs if they want to use functionality in Flux though? Seems like a rare edge case to only read the Zygote docs and the Flux optimizer docs without reading the Flux docs on taking gradients as well. I think updating the optimizer docs to add an example with `θ` would be a better fix than trying to shoehorn it in Zygote.

Reply:
I think it is confusing that two packages that are closely tied in practice (even if in reality Zygote exists independently of Flux) have contradictory statements. Flux explicitly encourages implicit parameters, and it makes no mention of the complex structure that results from `gradient(m -> loss(m(x), y), m)`. Zygote explicitly discourages implicit parameters. And the wording makes it sound like implicit and explicit work in Flux, but implicit is the "backwards compatible" version. But only certain varieties of explicit which are cumbersome to deal with for real models actually do work in Flux. The more convenient syntax with `m` will not.

Reply:
> And the wording makes it sound like implicit and explicit work in Flux, but implicit is the "backwards compatible" version

I haven't seen this verbiage anywhere, is it elsewhere in the Zygote docs? I certainly don't think putting implicit params second implies that their use is discouraged, but rather that it flows naturally along the gradient of easier -> more complex ideas. Note that the Flux docs use the same order, but actually endorse using implicit params for models with a lot of parameters (so pretty much all of them).

What I am against is explicitly and specifically calling out implicit params as the recommended way of doing things in the _Zygote_ docs. It's fine if we do so in the Flux docs (and we do), or if the language is neutral here (which IMO it is just after that sentence is removed).

Reply:
> However, implicit parameters exist mainly for compatibility with Flux's current AD; it's recommended to use the other approaches unless you need this.

In the last sentence of https://fluxml.ai/Zygote.jl/dev/#Gradients-of-ML-models-1

Reply:
I missed that, lets nix it in this PR or a follow-up.

Reply:
Maybe the solution is to beef up the last paragraph of this section, it's only a few sentences down. "implicit parameters exist mainly for compatibility with Flux's current AD;" sounds like it's a very temporary disclaimer, and "Flux-style model" makes it sounds like actual Flux works that way.

Maybe we delete "Flux-style model " and make the last paragraph something like this:

This implicit style is the one presently used by Flux, a closely related machine learning library. It uses structs like `Linear` above to define layers, and the function `Flux.params(model)` returns a `Params` object containing all the many parameters of its nested layers. For most other purposes, however, the explicit style is usually preferred.

Reply:
I like that!

Reply:
What the words still don't spell out is that "implicit parameters" means that the parameters are not arguments of the loss function. It takes no arguments, and picks up global (or, fairly global...) variables. 

I think you're supposed to conclude this from `() -> sum(linear(x))` and from "thread the parameter collection through" but I also think I was confused by that for a while. Maybe this paragraph/sentence should lay that out a bit more, like "Differentiate a zero-argument function, which closes over the parameters, and indicate that you want this behaviour by providing a Params object which exists for this purpose". Something like that?

Reply:
@JordiBolibar I believe we still want to remove this added sentence (in favor of @mcabbott's paragraph).

Reply:
I have made a small commit replacing it with @mcabbott 's paragraph, adding a short sentence before to make it flow better after the explicit section.